### PR TITLE
Add lock statement suport and block statement support for incremental parser

### PR DESCRIPTION
**Pull request by m36dot**
Add block statement and lock statement support for incremental parser.

<code>lock-stmt := lock block-stmt</code>
<code>sequence-stmt := statement*</code>
<code>block-stmt := { sequence-stmt }</code>

https://ballerina.io/spec/lang/2020R1/#lock-stmt

closes: #22486
closes: #22780

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 1]
```diff
@@ -72,7 +72,7 @@
             ParserRuleContext.WHILE_BLOCK, ParserRuleContext.CALL_STMT, ParserRuleContext.PANIC_STMT,
             ParserRuleContext.CONTINUE_STATEMENT, ParserRuleContext.BREAK_STATEMENT, ParserRuleContext.RETURN_STMT,
             ParserRuleContext.COMPOUND_ASSIGNMENT_STMT, ParserRuleContext.LOCAL_TYPE_DEFINITION_STMT,
-            ParserRuleContext.EXPRESSION_STATEMENT };
+            ParserRuleContext.EXPRESSION_STATEMENT, ParserRuleContext.LOCK_STMT };
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:75]: what about block stmt?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ */
+public class BlockStatement extends AbstractStatementTest {
+
+    // Valid syntax tests
+
+    @Test
+    public void testComplexBlockStmt() {
+        testFile("block-stmt/block_stmt_source_01.bal",
+        "block-stmt/block_stmt_assert_01.json");
+    }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:33]: This looks like a negative test case, but in the positive tests section?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    public void testBlockStmtWithMissingEqual() {
+        testFile("block-stmt/block_stmt_source_05.bal",
+        "block-stmt/block_stmt_assert_05.json");
+    }
+
+    @Test
+    public void testNestedObjectRecovery() {
+        testFile("block-stmt/block_stmt_source_06.bal",
+        "block-stmt/block_stmt_assert_06.json");
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:64]: This is a positive test case under recovery test section? 

Please check others as well, and lets re-arrange the stuff

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+}
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:67]: Any test for missing-closing parenthesis, etc?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/LockStatement.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    public void testLockStmtWithMissingEqual() {
+        testFile("lock-stmt/lock_stmt_source_05.bal",
+        "lock-stmt/lock_stmt_assert_05.json");
+    }
+
+    @Test
+    public void testNestedObjectRecovery() {
+        testFile("lock-stmt/lock_stmt_source_06.bal",
+        "lock-stmt/lock_stmt_assert_06.json");
+    }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/LockStatement.java:65]: Same comments I added to the block-stmt tests, applies to here as well.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 1 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:75]: sorry I missed that

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java (hunk 3 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:64]: this tests for nested recovery, for the missing semicolons and stuff inside the object type descritor

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java (hunk 4 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:67]: will add them :)

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java (hunk 3 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/BlockStatement.java:64]: oh ok, my bad.

**Pull request merged**